libecho/src/EchoProtocol.cc

```diff
@@ -0,0 +1,220 @@
+                buffer_send.push_back(reinterpret_cast<const uint8_t *>(buf)[pos]);
+                pos++;
+            }
+            while (buffer_send.size() == PACKET_SIZE) {
+                cv_send.wait(lock);
+            }
+        }
+        qDebug() << "write ending at position" << pos << "/" << count;
+        cv_send.notify_one();
+    }
+    qDebug() << "write successfully finished";
+    return pos;
+}
+
+size_t EchoProtocol::read(void *buf, size_t count, size_t timeout) {
+    size_t pos = 0;
+    std::unique_lock<std::mutex> lock(m_recv);
+    if (cv_recv.wait_for(lock, timeout * 1s, [&]{ return !buffer_recv.empty(); })) {
+        while (pos < count && !buffer_recv.empty()) {
+            reinterpret_cast<uint8_t *>(buf)[pos++] = buffer_recv.front();
+            buffer_recv.pop();
+        }
+        return pos;
+    }
+    throw std::runtime_error("connection broken");
+}
+
+void EchoProtocol::sendingThread(bool b) {
+    while (true) {
+        auto start = std::chrono::system_clock::now();
+        qDebug() << "Send start; status =" << status;
+        connection->sendStart();
+        if (status == 4) {
+            std::vector<uint8_t> vec = PacketBuilder().setFlag(Flag::FIN).getPacket().toBytes();
+            connection->send(vec.data(), vec.size());
+            qDebug() << "sent FIN";
+        } else if (status == 3) {
+            std::vector<uint8_t> vec = PacketBuilder().setFlag(Flag::DMD).getPacket().toBytes();
+            connection->send(vec.data(), vec.size());
+            qDebug() << "sent DMD";
+        } else {
+            if (b) {
+                lastPacket = PacketBuilder().setNumber(++number).setFlag(Flag::SYN).getPacket();
+                b = false;
+                qDebug() << "sent SYN, size" << lastPacket.getSize();
+            } else if (status == 0) {
+                if (closed && buffer_send.empty()) {
+                    thr[1]->detach();
+                    thr[1] = nullptr;
+                    close();
+                    return;
+                }
+                {
+                    std::unique_lock<std::mutex> lock(m_send);
+                    lastPacket = PacketBuilder().setData(buffer_send).setNumber(++number).getPacket();
+                    buffer_send.clear();
+                    qDebug() << "sent data, size" << lastPacket.getSize();
+                }
+                cv_send.notify_one();
+            } else if (status == 1) {
+                lastPacket = PacketBuilder().setNumber(number).setFlag(Flag::ACK1).getPacket();
+                qDebug() << "sent ACK, size" << lastPacket.getSize();
+                assert(!lastPacket.isSet(Flag::SYN));
+            } else {
+                qDebug() << "retransmitted last packet, size" << lastPacket.getSize();
+            }
+            std::vector<uint8_t> vec = lastPacket.toBytes();
+            auto debug = qDebug();
+            for (auto x: vec) {
+                debug << x;
+            }
```

Comment:
Czy qDebug nie umie printować vectorów tak po prostu? qDebug() << vec;
Jeśli nie, to może wygodniej by było dodać taką możliwość, niż tworzyć w kodzie takie pętle jak tu? 
https://doc.qt.io/qt-5/qdebug.html#writing-custom-types-to-a-stream